### PR TITLE
Add more typesafety for registered classes

### PR DESCRIPTION
**Pull request by AnonymousProgrammerLandau**
Crossing the border from C++ to Lua should keep type information for instances of registered classes:
 - [x] as parameter to Lua functions
 - [x] as return value of C++ callbacks

Obtaining references or pointers to instances of registered classes from Lua should fail if the actual type is not the requested:
 - [x] as parameter to C++ callbacks (raises an error in Lua)
 - [x] as return value of Lua functions
   - [x] for pointers (returns nullptr)
   - [x] for references (to be discussed #126)

This is a prerequisite to support C++ callbacks which return instances of registered classes by value, leaning on Lua's garbage collector to cleanup those values. That supports factory functions to emulate multiple constructors for classes without the need to invent an API to explicitly destroy those objects.

### CODE REVIEW

File: include/selene/Class.h [hunk 1]
```diff
@@ -87,7 +87,7 @@ class Class : public BaseClass {
     void _register_member(lua_State *state,
                           const char *fun_name,
                           Ret(T::*fun)(Args&&...)) {
-        std::function<Ret(T*, Args&&...)> lambda = [fun](T *t, Args&&... args) {
+        std::function<Ret(T*, Args&&...)> lambda = [fun](T *t, Args&&... args) -> Ret {
```
Comment (AnonymousProgrammerLandau) [inline include/selene/Class.h]: Enable references as return types.

File: include/selene/Obj.h [hunk 2]
```diff
@@ -66,7 +66,7 @@ class Obj : public BaseObj {
                           T *t,
                           const char *fun_name,
                           Ret(T::*fun)(Args&&...)) {
-        std::function<Ret(Args&&...)> lambda = [t, fun](Args&&... args) {
+        std::function<Ret(Args&&...)> lambda = [t, fun](Args&&... args) -> Ret {
```
Comment (AnonymousProgrammerLandau) [inline include/selene/Obj.h]: Enable references as return types.

**Comment by jeremyong** [comment]
This is looking good and addresses an important point. It's making me wonder if a lot of the functions would have benefitted from being private member functions to have access to the metatable registry as a private member but the way you've implemented it is fine

**Comment by AnonymousProgrammerLandau** [comment]
I also thought about private member functions to give access to the metatable registry as a private member. It seemed a more invasive change to me, especially considering the overloads for `sel::function`.

File: include/selene/BaseFun.h [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 #include "exotics.h"
 #include <exception>
-#include "exception.h"
+#include "ExceptionHandler.h"
```
Comment (AnonymousProgrammerLandau) [inline include/selene/BaseFun.h]: Renamed `exception.h` to `ExceptionHandler.h` to emphasize the difference between the handling of internally thrown exceptions and the newly user visible exception types in `ExceptionTypes.h`

**Comment by AnonymousProgrammerLandau** [comment]
@jeremyong: I have no further additions for this. Squash again?

**Comment by jeremyong** [comment]
Yes please!

**Pull request merged**